File: day16/puzzle16b.nim
import
    strutils,
    strformat,
    sequtils,
    sugar


const inputFileName = "input.txt"

type
    Message = object
        typeId: int
        value: int
        version: int
        versionSum: int
        messages: seq[Message]

proc parseOperatorMessage(message: string, startPos: int, typeId: int): (Message, int)

# -----------------------------------------------------------------------------
proc readInput(): string =
    let inputFile = open(inputFileName)
    defer: close(inputFile)

    return inputFile.readAll.strip

# -----------------------------------------------------------------------------
proc parseLiteralValue(packet: string, startPos: int, message: Message): (Message, int) =
    var
        pos: int = startPos
        valueString: string = ""
        msg = message

    while true:
        let more = packet[pos]
        pos += 1
        valueString &= packet[pos ..< pos + 4]
        pos += 4

        if $more == "0":
            break

    msg.value = fromBin[int](valueString)
    
    return (msg, pos)

# -----------------------------------------------------------------------------
proc parseMessage(message: string, startPos: int): (Message, int) =
    var
        pos: int = startPos
        typeId: int
        msg: Message

    if startPos >= len(message):
        return (msg, pos)

    let version = fromBin[int](message[pos .. pos + 2])
    pos += 3

    typeId = fromBin[int](message[pos .. pos + 2])
    pos += 3

    msg.typeId = typeId
    msg.version = version
    
    if typeId == 4:
        msg.typeId = typeId
        msg.version = version
        (msg, pos) = parseLiteralValue(message, pos, msg)
    else:
        (msg, pos) = parseOperatorMessage(message, pos, typeId)
        msg.version = version
        msg.typeId = typeId
    
    return (msg, pos)

# -----------------------------------------------------------------------------
proc parseOperatorMessage(message: string, startPos: int, typeId: int): (Message, int) =
    var
        pos: int = startPos
        length: int
        msg: Message
        subMsg: Message

    msg.typeId = typeId
    msg.value = 0
    msg.version = 0
    msg.messages = @[]

    let lengthTypeId = $message[pos]
    pos += 1

    case lengthTypeId
        of "0":
            
            # This means the length is represented by 15 bits, and the total length
            # of the rest of the message is the value of the length.
            length = fromBin[int](message[pos ..< pos + 15])
            pos += 15
            let endPos = pos + length - 6

            while pos < endPos:
                # Read the sub packets
                (subMsg, pos) = parseMessage(message, pos)
                msg.versionSum += subMsg.version
                msg.messages.add(subMsg)

        of "1":
            # This means the length is represented by 11 bits, and the length represents
            # the number of sub-packets
            length = fromBin[int](message[pos ..< pos + 11])
            pos += 11
            
            for i in 1 .. length:
                (subMsg, pos) = parseMessage(message, pos)
                msg.versionSum += subMsg.version
                msg.messages.add(subMsg)
        else:
            echo &"    Invalid length type id {lengthTypeId}"
    
    case msg.typeId
        of 0:     # Sum packets
            msg.value = msg.messages.map(m => m.value).foldl(a + b)

        of 1:    # Multiply packets
            msg.value = 1

            for subMsg in msg.messages:
                msg.value *= subMsg.value

        of 2:    # Get minimum of packets
            msg.value = msg.messages.map(m => m.value).foldl(if a < b: a else: b)

        of 3:    # Get maximum of packets
            msg.value = msg.messages.map(m => m.value).foldl(if a > b: a else: b)

        of 5:    # Greater than packets
            msg.value = if msg.messages[0].value > msg.messages[1].value: 1 else: 0

        of 6:    # Less than packets
            msg.value = if msg.messages[0].value < msg.messages[1].value: 1 else: 0

        of 7:    # Equal to packets
            msg.value = if msg.messages[0].value == msg.messages[1].value: 1 else: 0

        else:
            echo(&"Invalid message type id {msg.typeId}")

    return (msg, pos)

# -----------------------------------------------------------------------------
proc sumVersions(messages: seq[Message]): int =
    var versionSum = 0

    for msg in messages:
        versionSum += msg.version
        
        if len(msg.messages) > 0:
            versionSum += sumVersions(msg.messages)

    return versionSum

# -----------------------------------------------------------------------------
proc parseMessages() =
    var
        position: int = 0
        messages: seq[Message] = @[]
        msg: Message

    let packets = readInput()

    # Convert hex string to binary
    let bin = packets
        .map(c => $c)
        .map(c => fromHex[int]($c))
        .map(h => toBin(h, 4))
        .join("")

    # Parse the packets of the binary message packet
    while true:
        let version = fromBin[int](bin[position .. position + 2])
        position += 3

        let typeId = fromBin[int](bin[position .. position + 2])
        position += 3

        msg.version = version
        msg.typeId = typeId

        if typeId == 4:
            (msg, position) = parseLiteralValue(bin, position, msg)
            messages.add(msg)
        else:
            (msg, position) = parseOperatorMessage(bin, position, typeId)

        messages.add(msg)

        if position mod 4 != 0:
            position += 4 - (position mod 4)

        if position >= len(bin) - 6:
            break

    let versionSum = sumVersions(messages)
    echo &"Version Sum: {versionSum}"
    echo &"Value: {msg.value}"

# -----------------------------------------------------------------------------

when isMainModule:
    parseMessages()

File: day16/puzzle16b.py
from dataclasses import dataclass
import dataclasses
from pathlib import Path

inputFileName = "input.txt"

@dataclass
class Message:
    typeId: int
    value: int
    version: int
    versionSum: int = 0
    messages = []

# -----------------------------------------------------------------------------
def readInput() -> str:
    cwd = Path(__file__).parent.absolute()

    with open(cwd / inputFileName) as f:
        return f.read().strip()

# -----------------------------------------------------------------------------
def parseLiteralValue(packets: str, startPos: int, message: Message):
    pos: int = startPos
    valueString: str = ""

    while True:
        more = packets[pos]
        pos += 1
        valueString += packets[pos:pos + 4]
        pos += 4

        if more == "0":
            break

    message.value = int(valueString, 2)

    return message, pos

# -----------------------------------------------------------------------------
def parseMessage(message: str, startPos: int):
    pos: int = startPos
    typeId: int = 0

    if startPos >= len(message):
        return None, pos

    version = int(message[pos: pos + 3], 2)
    pos += 3

    typeId = int(message[pos: pos + 3], 2)
    pos += 3

    if typeId == 4:
        msg = Message(typeId, 0, version)
        msg, pos = parseLiteralValue(message, pos, msg)
    else:
        msg, pos = parseOperatorMessage(message, pos, typeId)
        msg.version = version
        msg.typeId = typeId

    return msg, pos

# -----------------------------------------------------------------------------
def parseOperatorMessage(message: str, startPos: int, typeId: int):
    pos: int = startPos
    length: int

    msg = Message(typeId, 0, 0)
    msg.messages = []
    lengthTypeId = message[pos]
    pos += 1

    match lengthTypeId:
        case "0":

            # This means the length is represented by 15 bits, and the total length
            # of the rest of the message is the value of the length.
            length = int(message[pos: pos + 15], 2)
            pos += 15
            endPos = pos + length

            while pos < endPos - 6:
                # Read the sub packets
                subMsg, pos = parseMessage(message, pos)
                msg.versionSum += subMsg.version
                msg.messages.append(subMsg)

        case "1":
            # This means the length is represented by 11 bits, and the length represents
            # the number of sub-packets
            length = int(message[pos: pos + 11], 2)
            pos += 11

            for i in range(length):
                subMsg, pos = parseMessage(message, pos)
                msg.versionSum += subMsg.version
                msg.messages.append(subMsg)

        case _:
            print(f"    Invalid length type id {lengthTypeId}")

    match msg.typeId:
        case 0:     # Sum packets
            msg.value = sumValues(msg.messages)

        case 1:    # Multiply packets
            msg.value = 1

            for subMsg in msg.messages:
                msg.value *= subMsg.value

        case 2:    # Get minimum of packets
            msg.value = min(map(lambda x: x.value, msg.messages))

        case 3:    # Get maximum of packets
            msg.value = max(map(lambda x: x.value, msg.messages))

        case 5:    # Greater than packets
            msg.value = 1 if msg.messages[0].value > msg.messages[1].value else 0

        case 6:    # Less than packets
            msg.value = 1 if msg.messages[0].value < msg.messages[1].value else 0

        case 7:    # Equal to packets
            msg.value = 1 if msg.messages[0].value == msg.messages[1].value else 0

        case _:
            print(f"    Invalid message type id {msg.typeId}")

    return msg, pos

# -----------------------------------------------------------------------------
def parseMessages():
    position = 0
    messages: list[Message] = []

    packets = readInput()

    # Convert hex string to binary
    bin = "".join([f"{int(c, 16):04b}" for c in packets])

    # Parse the packets of the binary message packet
    while True:
        version = int(bin[position: position + 3], 2)
        position += 3

        typeId = int(bin[position: position + 3], 2)
        position += 3

        if typeId == 4:
            msg, position = parseLiteralValue(bin, 6)
            messages.append(msg)
        else:
            msg, position = parseOperatorMessage(bin, position, typeId)
            msg.typeId = typeId
            msg.version = version

        msg.version = version
        messages.append(msg)

        if position % 4 != 0:
            position += 4 - (position % 4)

        if position >= len(bin) - 6:
            break

    print(f"Value: {msg.value}")
   
# -----------------------------------------------------------------------------
def sumVersions(messages: list[Message]):
    versionSum = 0

    for msg in messages:
        versionSum += msg.version

        if len(msg.messages) > 0:
            versionSum += sumVersions(msg.messages)

    return versionSum

# -----------------------------------------------------------------------------
def sumValues(messages: list[Message]):
    sum = 0

    for msg in messages:
        sum += msg.value

        # if len(msg.messages) > 0:
        #     sum += sumValues(msg.messages)

    return sum

# -----------------------------------------------------------------------------
if __name__ == "__main__":
    parseMessages()
